python/src/scipp/plot/resampling_model.py

```diff
@@ -0,0 +1,198 @@
+                        out = out[sc.get_slice_params(out.data, out.meta[dim],
+                                                      low, high)]
+                params[dim] = (low.value, high.value, low.unit,
+                               self.resolution[dim])
+        if params == self._home_params:
+            # This is a crude caching mechanism for past views. Currently we
+            # have the "back" buttons disabled in the matplotlib toolbar, so
+            # we cache only the "home" view. This is the most expensive to
+            # create anyway.
+            self._resampled_params = self._home_params
+            self._resampled = self._home
+        elif self._resampled is None or params != self._resampled_params:
+            self._resampled_params = params
+            self._edges = self._make_edges(params)
+            self._resampled = self._resample(out)
+        if self._home is None:
+            self._home = self._resampled
+            self._home_params = self._resampled_params
+
+
+class ResamplingBinnedModel(ResamplingModel):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # TODO See #1469. This is a temporary hack to work around the
+        # conversion of coords to edges in model.py.
+        for name in self._array.meta:
+            if name == 'spectrum':
+                self._array.meta['spectrum'] = 0.5 * (
+                    self._array.meta['spectrum']['spectrum', 1:] +
+                    self._array.meta['spectrum']['spectrum', :-1])
+
+    def _resample(self, array):
+        # We could bin with all edges and then use `bins.sum()` but especially
+        # for inputs with many bins handling the final edges using `histogram`
+        # is faster with the current implementation of `sc.bin`.
+        binned = sc.bin_with_coords(array.data, array.meta, self.edges[:-1],
+                                    [])
+        a = sc.histogram(binned, self.edges[-1])
+        for name, mask in array.masks.items():
+            a.masks[name] = self._rebin(mask, array.meta)
+        return a
+
+
+class ResamplingCountsModel(ResamplingModel):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _resample(self, array):
+        return sc.DataArray(
+            data=self._rebin(array.data, array.meta),
+            coords={edge.dims[-1]: edge
+                    for edge in self.edges},
+            masks={
+                name: self._rebin(mask, array.meta)
+                for name, mask in array.masks.items()
+            })
+
+
+class ResamplingDenseModel(ResamplingModel):
+    def __init__(self, array, **kwargs):
+        # Scale by bin widths, so `rebin` is effectively performing a "mean"
+        # operation instead of "sum".
+        super().__init__(self._to_density(array), **kwargs)
+
+    def _to_density(self, array):
+        array = PlotArray(data=array.data.astype(sc.dtype.float64),
```

Comment:
Is `astype` here always making a copy? Isn't that potentially expensive?

Reply:
See our discussion from 10 days ago: We *have* to make a copy because we have to rescale before rebin. Since this is only used for non-count data we currently to not have many performance-critical applications for this, I believe.

Reply:
yes but in master, we are making a copy of at most a 2D array, because the copy was made after having sliced away the additional dimensions. Here, i think you are copying the entire ND data, right?

Reply:
True, but this would disappear once we support rebinning non-counts data in one way or another.